erc777.k

```diff
@@ -69,140 +72,379 @@ module ERC777
+        ~> #atomicEnd(#void) ... </k>
+       <caller> Operator </caller>
+       <id> Holder </id>
+       <balance> Balance </balance>
+
+  syntax Stmt ::= #burn ( Address , Number )
+
+  rule <k> #burn(Holder, Amount) => . ... </k>
+       <account>
+         <id> Holder </id>
+         <balance> Balance => Balance -Int Amount </balance>
+         ...
+       </account>
+       <supply> Supply => Supply -Int Amount </supply>
+       requires Balance >=Int Amount andBool Supply >=Int Amount
+
+  // general version
+  syntax Stmt ::= #log ( Event )
+                | #logERC20 ( Event )
+
+  rule <k> #log(Log) => . ... </k>
+       <log> Logs => Logs Log </log>
+
+  rule <k> #logERC20(Log) => . ... </k>
+       <log> Logs => Logs Log </log>
+       <ERC20CompatibleC> ERC20Compatible </ERC20CompatibleC>
+       requires ERC20Compatible
+
+  rule <k> #logERC20(Log) => . ... </k>
+       <ERC20CompatibleC> ERC20Compatible </ERC20CompatibleC>
+       requires notBool ERC20Compatible
+
+
+  //
+  // mint -> no function spec, but rules
+  // TODO: `event Minted` signature inconsistency in standard
 
-  rule #isMultipleOf ( Value , Granularity ) => Value /Int Granularity *Int Granularity ==Int Value
+  syntax Stmt ::= #mint ( Address , Number , Data , Data )
+                | #mintAux ( Address , Number )
 
-  syntax Stmt ::= #call ( Address , Address , Address , Address , Address , Number , Data , Data )
+  rule <k> #mint(Holder, Amount, Data, OperatorData)
+        => #atomicBegin
+        ~> #ensure(#and(Amount >=Int 0, #isMultipleOf(Balance +Int Amount, granularity()),
+                        Holder =/=Int 0))
+        ~> #mintAux(Holder, Amount)
+        ~> #callRecipent(Operator, 0, Holder, Amount, Data,  OperatorData, true)
+        ~> #log(Minted(Operator, Holder, Amount, Data, OperatorData))
+        ~> #logERC20(Transfer(0, Holder, Amount))
+        ~> #atomicEnd(#void) ... </k>
+       <caller> Operator </caller>
+       <id> Holder </id>
+       <balance> Balance </balance>
+
+  rule <k> #mintAux(Holder, Amount) => . ... </k>
+       <account>
+         <id> Holder </id>
+         <balance> Balance => Balance +Int Amount </balance>
+         ...
+       </account>
+       <supply> Supply => Supply +Int Amount </supply>
+       requires Balance +Int Amount <=Int #MAXVALUE
+
+  // Hooks
+  // TODO: 
+  // * check if registered via ERC820?
+  // * check if hook call can alter token contract state
+
+  rule <k> Hook . tokensToSend ( _ , _ , _ , _ , _ , _ ) => . ... </k>
+       <id> Hook </id>
+       <hookFail> false </hookFail>
+
+  rule <k> Hook . tokensToSend ( _ , _ , _ , _ , _ , _ ) => revert ... </k>
+       <id> Hook </id>
+       <hookFail> true </hookFail>
+
+  rule <k> Hook . tokensReceived ( _ , _ , _ , _ , _ , _ ) => . ... </k>
+       <id> Hook </id>
+       <hookFail> false </hookFail>
+
+  rule <k> Hook . tokensReceived ( _ , _ , _ , _ , _ , _ ) => revert ... </k>
+       <id> Hook </id>
+       <hookFail> true </hookFail>
```

Comment:
i'm not sure if the reader will get the idea that `<hookFail>` will be updated externally asynchronously.